runtime/src/registrar.rs

```diff
@@ -0,0 +1,1346 @@
+		RetryQueue::mutate(|q| {
+			for (i, q) in q.iter_mut().enumerate() {
+				if !q.is_empty() {
+					return Some((q.remove(0), i as u32));
+				}
+			}
+			None
+		})
+	}
+
+	/// Forcibly remove the threads matching `m` from all current and future scheduling.
+	fn force_unschedule(m: impl Fn(ParaId) -> bool) {
+		RetryQueue::mutate(|qs| for q in qs.iter_mut() {
+			q.retain(|i| !m(i.0))
+		});
+		SelectedThreads::mutate(|qs| for q in qs.iter_mut() {
+			q.retain(|i| !m(i.0))
+		});
+		Active::mutate(|a| for i in a.iter_mut() {
+			if m(i.0) {
+				if let Some((_, ref mut r)) = i.1 {
+					*r = Retriable::Never;
+				}
+			}
+		});
+	}
+}
+
+impl<T: Trait> ActiveParas for Module<T> {
+	fn active_paras() -> Vec<(ParaId, Option<(CollatorId, Retriable)>)> {
+		Active::get()
+	}
+}
+
+/// Ensure that parathread selections happen prioritized by fees.
+#[derive(Encode, Decode, Clone, Eq, PartialEq)]
+pub struct LimitParathreadCommits<T: Trait + Send + Sync>(rstd::marker::PhantomData<T>) where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>;
+
+#[cfg(feature = "std")]
+impl<T: Trait + Send + Sync> rstd::fmt::Debug for LimitParathreadCommits<T> where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>
+{
+	fn fmt(&self, f: &mut rstd::fmt::Formatter) -> rstd::fmt::Result {
+		write!(f, "LimitParathreadCommits<T>")
+	}
+}
+
+/// Custom validity errors used in Polkadot while validating transactions.
+#[repr(u8)]
+pub enum Error {
+	/// Parathread ID has already been submitted for this block.
+	Duplicate = 0,
+	/// Parathread ID does not identify a parathread.
+	InvalidId = 1,
+}
+
+impl<T: Trait + Send + Sync> SignedExtension for LimitParathreadCommits<T> where
+	<T as system::Trait>::Call: IsSubType<Module<T>, T>
+{
+	type AccountId = T::AccountId;
+	type Call = <T as system::Trait>::Call;
+	type AdditionalSigned = ();
+	type Pre = ();
+
+	fn additional_signed(&self)
+		-> rstd::result::Result<Self::AdditionalSigned, TransactionValidityError>
+	{
+		Ok(())
+	}
+
+	fn validate(
+		&self,
+		_who: &Self::AccountId,
+		call: &Self::Call,
+		_info: DispatchInfo,
+		_len: usize,
+	) -> TransactionValidity {
+		let mut r = ValidTransaction::default();
+		if let Some(local_call) = call.is_sub_type() {
+			if let Call::select_parathread(id, collator, hash) = local_call {
+				// ensure that the para ID is actually a parathread.
+				let e = TransactionValidityError::from(InvalidTransaction::Custom(Error::InvalidId as u8));
+				<Module<T>>::ensure_thread_id(*id).ok_or(e)?;
+
+				// ensure that we haven't already had a full complement of selected parathreads.
+				let mut upcoming_selected_threads = SelectedThreads::get();
+				if upcoming_selected_threads.is_empty() {
+					upcoming_selected_threads.push(vec![]);
+				}
+				let i = upcoming_selected_threads.len() - 1;
+				let selected_threads = &mut upcoming_selected_threads[i];
+				let thread_count = ThreadCount::get() as usize;
+				ensure!(
+					selected_threads.len() < thread_count,
```

Comment:
since we've already pushed onto `selected_threads` at this point, could it be `selected_threads.len() <= thread_count`?